File: sai-rn-agent-src/android/app/src/main/java/com/saicompanion/SaiDeviceControlModule.java
package com.saicompanion;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import android.content.Intent;
import android.provider.Settings;

public class SaiDeviceControlModule extends ReactContextBaseJavaModule {
    private final ReactApplicationContext reactContext;
    private SaiHttpServer server;

    public SaiDeviceControlModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "SaiDeviceControl";
    }

    @ReactMethod
    public void startLocalServer(int port, Promise promise) {
        try {
            if (server == null) {
                server = new SaiHttpServer(port, reactContext);
                server.start();
            }
            promise.resolve("Server started on port " + port);
        } catch (Exception e) {
            promise.reject("SERVER_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void stopLocalServer(Promise promise) {
        if (server != null) {
            server.stop();
            server = null;
        }
        promise.resolve("Server stopped");
    }

    @ReactMethod
    public void checkAccessibilityPermission(Promise promise) {
        // Logic to check if Accessibility is enabled
        promise.resolve(true); // Simplified for example
    }

    @ReactMethod
    public void openAccessibilitySettings() {
        Intent intent = new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        reactContext.startActivity(intent);
    }
}

File: sai-rn-agent/android/app/src/main/java/com/saicompanion/SaiDeviceControlModule.kt
package com.saicompanion

import android.content.Intent
import android.provider.Settings
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class SaiDeviceControlModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var server: SaiHttpServer? = null

    override fun getName(): String {
        return "SaiDeviceControl"
    }

    @ReactMethod
    fun startLocalServer(port: Int, promise: Promise) {
        try {
            if (server == null) {
                server = SaiHttpServer(port, reactContext)
                server?.start()
            }
            promise.resolve("Server started on port $port")
        } catch (e: Exception) {
            promise.reject("SERVER_ERROR", e.message)
        }
    }

    @ReactMethod
    fun stopLocalServer(promise: Promise) {
        if (server != null) {
            server?.stop()
            server = null
        }
        promise.resolve("Server stopped")
    }

    @ReactMethod
    fun checkAccessibilityPermission(promise: Promise) {
        // Here you would optimally add logic checking Settings.Secure.ENABLED_ACCESSIBILITY_SERVICES
        val isEnabled = SaiAccessibilityService.instance != null
        promise.resolve(isEnabled)
    }

    @ReactMethod
    fun openAccessibilitySettings() {
        val intent = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        reactContext.startActivity(intent)
    }
}
